src/System.Private.CoreLib/shared/System/SpanHelpers.Byte.cs

```diff
@@ -196,107 +195,198 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
+                    {
+                        Vector128<byte> search = LoadVector128(ref searchSpace, offset);
+
+                        int matches = Sse2.MoveMask(Sse2.CompareEqual(values, search));
+                        if (matches == 0)
+                        {
+                            offset += Vector128<byte>.Count;
+                            continue;
+                        }
+
+                        // Find offset of first match
+                        return ((int)(byte*)offset) + BitOps.TrailingZeroCount(matches);
+                    }
+
+                    if ((int)(byte*)offset < length)
+                    {
+                        nLength = (IntPtr)(length - (int)(byte*)offset);
+                        goto SequentialScan;
+                    }
+                }
+            }
+            else if (Vector.IsHardwareAccelerated)
+            {
+                if ((int)(byte*)offset < length)
                 {
-                    nLength = (IntPtr)(length - (int)(byte*)index);
-                    goto SequentialScan;
+                    nLength = GetByteVectorSpanLength(offset, length);
+
+                    Vector<byte> values = new Vector<byte>(value);
+
+                    while ((byte*)nLength > (byte*)offset)
+                    {
+                        var matches = Vector.Equals(values, LoadVector(ref searchSpace, offset));
+                        if (Vector<byte>.Zero.Equals(matches))
+                        {
+                            offset += Vector<byte>.Count;
+                            continue;
+                        }
+
+                        // Find offset of first match
+                        return (int)(byte*)offset + LocateFirstFoundByte(matches);
+                    }
+
+                    if ((int)(byte*)offset < length)
+                    {
+                        nLength = (IntPtr)(length - (int)(byte*)offset);
+                        goto SequentialScan;
+                    }
                 }
             }
             return -1;
         Found: // Workaround for https://github.com/dotnet/coreclr/issues/13549
-            return (int)(byte*)index;
+            return (int)(byte*)offset;
         Found1:
-            return (int)(byte*)(index + 1);
+            return (int)(byte*)(offset + 1);
         Found2:
-            return (int)(byte*)(index + 2);
+            return (int)(byte*)(offset + 2);
         Found3:
-            return (int)(byte*)(index + 3);
+            return (int)(byte*)(offset + 3);
         Found4:
-            return (int)(byte*)(index + 4);
+            return (int)(byte*)(offset + 4);
         Found5:
-            return (int)(byte*)(index + 5);
+            return (int)(byte*)(offset + 5);
         Found6:
-            return (int)(byte*)(index + 6);
+            return (int)(byte*)(offset + 6);
         Found7:
-            return (int)(byte*)(index + 7);
+            return (int)(byte*)(offset + 7);
```

Comment:
Can we use a variable to avoid the multiple goto targets? For example
```csharp

if (uValue == Unsafe.AddByteOffset(ref searchSpace, offset + i))
{
    delta = i;
    goto Found;
}

...

Found: 
    return (int)(byte*)(offset + delta);
```

Reply:
Not sure, pushes more code into the fast chain (running through the compares) to avoid more in the slow area (the targets)?

Currently these sections looks like this:
```asm
movzx    r11, byte  ptr [rcx+r11]
cmp      r11d, eax
je       G_M10207_IG14
lea      r11, [r9+2]
movzx    r11, byte  ptr [rcx+r11]
cmp      r11d, eax
je       G_M10207_IG15
lea      r11, [r9+3]
movzx    r11, byte  ptr [rcx+r11]
cmp      r11d, eax
je       G_M10207_IG16
lea      r11, [r9+4]
movzx    r11, byte  ptr [rcx+r11]
cmp      r11d, eax
je       G_M10207_IG17
lea      r11, [r9+5]
movzx    r11, byte  ptr [rcx+r11]
cmp      r11d, eax
je       G_M10207_IG18
lea      r11, [r9+6]
movzx    r11, byte  ptr [rcx+r11]
cmp      r11d, eax
je       G_M10207_IG19
lea      r11, [r9+7]
movzx    r11, byte  ptr [rcx+r11]
cmp      r11d, eax
je       G_M10207_IG20
```

Reply:
> pushes more code into the fast chain (running through the compares) to avoid more in the slow area (the targets)?

Hmm, I always prefer smaller code-size over the "tricky" loop-unrolling (if the loop-body does not have long latency instructions). Sometimes, loop-unrolling can have a bit more beautiful perf data on microbenchmarks, but I believe I-cache is the most precious resource for the large real applications (e.g., ASP.NET servers).

Additionally, if a loop body is small enough, that would be specially optimized on Intel architectures (please see "Loop Stream Detector" sections in Intel optimization manual). I think most of the fast chain loops could be really small.

Reply:
For completeness the exit points are
```asm
G_M10207_IG11:
    mov      eax, -1
G_M10207_IG12:
   vzeroupper 
   ret  
G_M10207_IG13:
   mov      eax, r9d
   jmp      SHORT G_M10207_IG21
G_M10207_IG14:
   lea      rax, [r9+1]
   jmp      SHORT G_M10207_IG21
G_M10207_IG15:
   lea      rax, [r9+2]
   jmp      SHORT G_M10207_IG21
G_M10207_IG16:
   lea      rax, [r9+3]
   jmp      SHORT G_M10207_IG21
G_M10207_IG17:
   lea      rax, [r9+4]
   jmp      SHORT G_M10207_IG21
G_M10207_IG18:
   lea      rax, [r9+5]
   jmp      SHORT G_M10207_IG21
G_M10207_IG19:
   lea      rax, [r9+6]
   SHORT G_M10207_IG21
G_M10207_IG20:
   lea      rax, [r9+7]
G_M10207_IG21:
   vzeroupper 
   ret      
```

Reply:
A follow up turning them to regular `for` loops might be worth investigating then?

Reply:
The current micro-benchmark oriented performance culture in the .NET Core repos favors bigger streamlined code because of it tends to produce better results in microbenchmarks. It is a common wisdom (at Microsoft at least) that smaller code runs faster in real workloads because of the factors @fiigii mentioned. The code bloating optimizations are worth it in like 1% of the cases. You have seen me pushing back on the more extreme cases of code bloat. This case is on the edge. My gut-feel is that this one is probably good as it is. We would need to be able to get data about performance in real workloads to tell with confidence.

Reply:
> A follow up turning them to regular for loops might be worth investigating then?

Yes, that should be a follow-up work, not in this PR.

Reply:
Definitely worth revisiting; lots of remarks that amount to the CPU might not be happy with 8 conditional branches in quick succession e.g.

> Assembly/Compiler Coding Rule 10. (M impact, L generality) Do not put more than four
branches in a 16-byte chunk.

Reply:
Though not sure how you could; as here the jump are 6 bytes and the compares are 4 bytes... be a push to get 4 compares and 4 jumps in 16 bytes?